Return fewer change outputs when the minimum ada quantity cannot be satisfied.

jonathanknowles opened the pull request:
# Issue Number

https://jira.iohk.io/projects/ADP/issues/ADP-696

# Overview

By default, the coin selection algorithm attempts to make exactly one change output for each user-specified output.

But in situations where the supply of ada is limited, this strategy can lead to failure if it is not possible to assign the minimum ada quantities to all change outputs.

This PR relaxes this restriction, so that in situations where ada is limited, we allow the algorithm to return fewer change outputs. 

# Comments

- [x] Adjust coin selection algorithm to allow a reduced number of change outputs when the supply of ada is limited.
- [x] Update property tests to allow a slightly-increased fee in the case when change outputs have been dropped.
- [x] Revises the way we sort change maps within `makeChange` so that:
    - Empty change bundles are placed at the start.
    - Empty change bundles deriving from **_user-specified assets_** and empty change bundles deriving from **_non-user-specified assets_** are combined together, if possible.
- [x] Add property test to check that sorting token bundles by `AssetCount` always places coins before non-coins.
- [ ] Add property test to check that it is possible to send the entire balance to another wallet.
    - [ ] Generalize this property to work with non-zero minimum ada quantities.

jonathanknowles committed:
Add newline characters before complex counterexample values.

This makes counterexamples much easier to read in the case where there are
complex values involved (which already span multiple lines).

jonathanknowles committed:
Minimize the central monadic expression of `makeChange`.

There are two sets of values in this function that are completely
determined by the function's arguments.

Before this change, some were listed in the `where` clause, but others
were listed in the `let` clause of the function body.

This change simply consolidates them into a single place.

As a result, the function's main monadic expression is minimized.

jonathanknowles committed:
Remove arguments to `changeError`.

These values are already completely determined by pre-defined values
within the function's scope.

jonathanknowles committed:
Make it possible for `makeChange` to return an empty list.

jonathanknowles committed:
Rename `extraInputCoins` to `extraInputCoin`.

There can only ever be one coin here.

jonathanknowles committed:
Rename `availableCoins` to `availableCoin`.

There can only ever be one coin passed into this function.

(The singular coin value will of course get smaller through repeated
application of this function to token maps in a sequence.)

jonathanknowles committed:
Rename `change` to `changeForAssets`.

This makes it a bit clearer that this particular set of change is for
non-ada assets only (both user-specified and non-user-specified).

jonathanknowles committed:
Rename `bundles` to `changeForAssetsWithMinimalCoins`.

This makes it clearer that the minimum ada amount has already been
assigned to these change bundles.

jonathanknowles committed:
Rename `remainder` to `leftoverCoin`.

This makes it a bit clearer that the leftover amount is a coin value, rather
than something else.

jonathanknowles committed:
Drop change outputs if the minimum ada quantity cannot be satisfied.

It is not possible to assign a minimum ada quantity to all change
outputs, then attempt to create a smaller number of change outputs.

This should only be necessary if the available supply of ada is limited.

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1288,11 +1288,15 @@ selectAssetsNoOutputs ctx wid tx transform = do
-    once fn (a :| as) = fn a :| as
+    once :: a -> (a -> a) -> [a] -> [a]
+    once _ f (a : as) = f a : as
+    once a _ [] = [a]
```
This needs checking.

rvl replied:
It seems a bit unnatural to have `changeGenerated = [mempty]`. Why not `[]`?

KtorZ replied:
This is wrong indeed. The name of the function was maybe confusing. This is what this piece of code is intended to do:

```hs
    reclaim `addToHead` changeGenerated sel
  where
    addToHead :: TokenBundle -> [TokenBundle] -> [TokenBundle]
    addToHead x []    = [x]
    addToHead x (h:q) = TokenBundle.add x h : q
```

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -508,41 +506,65 @@ performSelection minCoinValueFor costFor criteria
+                Nothing ->
+                    -- There is no more ada available. Terminate with a
+                    -- less-than-optimal number of change outputs.
+                    pure $ Right $ mkSelectionResult change
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -508,41 +506,65 @@ performSelection minCoinValueFor costFor criteria
+            { inputsSkeleton  = selected
+            , outputsSkeleton = NE.toList outputsToCover
+            , changeSkeleton
+            }
```
You really dislike let binding, don't you :smile:

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -791,105 +813,105 @@ makeChange
+        assignMinimumCoin minCoinFor <$> changeForAssets
+
+    totalMinimumCoinForAssets =
+        F.foldMap TokenBundle.getCoin changeForAssetsWithMinimalCoins
```
I find this a little confusing now to have these two declarations in the top-level `where` because they are only used for error reporting. That's why I originally put `totalMinCoinValue` under `changeError`. IMO, the nicest feature about `where` is that it can be nested to be of limited scope and local to where it is used, which helps a lot with readability. Having a top-level flat list of local bindings reduces readability.

jonathanknowles replied:
Hi @KtorZ Good point!

I originally moved this up to a higher scope as I wanted to reuse it from another place. However, I eventually changed my implementation so that this was no longer necessary.

I'm happy to move this back under the `where` clause (where it was before).

jonathanknowles replied:
Fixed in 3ca8552436fd1ccc23e1256f27e34032d7adbae6.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -914,6 +936,106 @@ makeChange minCoinValueFor requiredCost mExtraCoinSource inputBundles outputBund
+            -- bundles obtained by partitioning the remaining ada amount:
+            Just $ NE.toList $ NE.zipWith (<>)
+                bundlesForAssetsWithMinimumCoins
+                bundlesForOutputCoins
```
This case is the "normal" case as we had before.

KtorZ replied:
(just a self-note)

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -914,6 +936,106 @@ makeChange minCoinValueFor requiredCost mExtraCoinSource inputBundles outputBund
+assignCoinsToChangeMaps adaAvailable minCoinFor pairsAtStart =
+    loop adaRequiredAtStart pairsAtStart
+  where
+    loop !adaRequired !pairsNonEmpty = case pairsNonEmpty of
```
Isn't it a bit confusing that `adaRequired` comes as an argument here? I'd expect whatever branch needs it to re-calculate it in its context. That's the case for the first branch only.

jonathanknowles replied:
> Isn't it a bit confusing that adaRequired comes as an argument here?

We could re-calculate it on every iteration, but presumably that would make it necessary to compute the minimum values across all remaining token maps in the list, which would lead to quadratic run-time in the length of the list, unless you're thinking of something else?

Here I've opted to compute the total requirement up front, and then just reduce it on each iteration by the correct amount (which should be equal to the minimum ada value of the map that we drop).

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -914,6 +936,106 @@ makeChange minCoinValueFor requiredCost mExtraCoinSource inputBundles outputBund
+            -- asset map that is safe to drop. This will reduce the amount of
+            -- ada required by a small amount:
+            let adaRequired' = adaRequired `Coin.distance` minCoinFor m in
+            loop adaRequired' (p :| ps)
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -914,6 +936,106 @@ makeChange minCoinValueFor requiredCost mExtraCoinSource inputBundles outputBund
+            -- We didn't have any non-ada assets at all in our change. We can
+            -- simply return the available ada amount as a singleton change
+            -- bundle:
+            Just [TokenBundle.fromCoin adaAvailable]
```
Isn't that branch a special case of / redundant with the first branch? The first branch will always match before this one, and I'd indeed expect the result to be the same.

jonathanknowles replied:
I think you're right. I've fixed this in 427e208bfddbe07794927f2e3855761b530196cf.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -914,6 +936,106 @@ makeChange minCoinValueFor requiredCost mExtraCoinSource inputBundles outputBund
+            -- also don't have enough ada available to pay even for a single
+            -- change output. We just burn the available ada amount (which
+            -- will be small), returning no change.
+            Just []
```
Yes. I like that is guarded by `adaAvailable < adaRequired`. Although I'd really prefer that the `adaRequired` is actually computed within the function loop itself, and not provided as argument. It makes reasoning and following the steps much easier and it becomes clear that `adaRequired` is necessarily small.

jonathanknowles replied:
Hi @KtorZ 

I've written a justification for this in https://github.com/input-output-hk/cardano-wallet/pull/2490#discussion_r570169560. Would be interested to know your thoughts on that!

jonathanknowles committed:
Remove redundant pattern match in `assignCoinsToChangeMaps`.

This pattern match will never fire, since any value that matches it will
also match the first pattern match. The result should also be identical.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2490#discussion_r570159866

jonathanknowles committed:
Revise the way we sort change within `makeChange`.

The new sorting method arranges that empty change maps are located at
the start of the list of all change maps, so that when we eventually
call `assignCoinsToChangeMaps`, it has a greater chance of succeeding.

The correspondence between output coins and change for user-specified
assets is preserved.

jonathanknowles committed:
Calculate `totalMinCoinValue` within scope of `changeError`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2490#discussion_r570156188

jonathanknowles committed:
Adjust `TransactionSpec` to use empty lists for change outputs.

jonathanknowles committed:
Rename `knownAssetIds` to `userSpecifiedAssetIds`.

This is consistent with the naming elsewhere.

jonathanknowles committed:
Add documentation to `userSpecifiedAssets` and `nonUserSpecifiedAssets`.

rvl reviewed (commented):
Looks pretty good so far!

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:
```diff
@@ -533,33 +579,52 @@ prop_performSelection minCoinValueFor costFor (Blind criteria) coverage =
-        let delta = TokenBundle.unsafeSubtract
-                balanceSelected
-                (balanceRequired <> balanceChange)
+        assert True
```
:thinking: ?

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:
```diff
@@ -533,33 +579,52 @@ prop_performSelection minCoinValueFor costFor (Blind criteria) coverage =
             balanceSelected (balanceRequired <> balanceChange)
-        assert $ TokenBundle.getCoin delta == expectedCost
+        assert $ TokenBundle.getCoin delta >= expectedCost
+        assert $ TokenBundle.getCoin delta <= maximumExpectedDelta
```
:+1:

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:
```diff
@@ -533,33 +579,52 @@ prop_performSelection minCoinValueFor costFor (Blind criteria) coverage =
+        maximumExpectedDelta =
+            expectedCost `addCoin`
+            (absoluteMinCoinValue `multiplyCoin`
+                (length outputsCovered - length changeGenerated))
```
I'd expect the maximum delta to actually be: `expectedCost + absoluteMinCoinValue` no matter what. Did QC find counter-examples of that already :thinking: ?

KtorZ replied:
After checking and fiddling around with that, there are indeed cases where may not be able to create _any_ change output, in which case this property is indeed verified.

KtorZ reviewed (approved):
Looks good. Thanks Jonathan :+1:

KtorZ committed:
add some unit tests for 'assignCoinsToChangeMaps' as sanity checks

KtorZ committed:
add some extra counter-examples in prop_performSelection to help debugging

KtorZ committed:
fix surplus assignation when re-constructing selection for delegation/undelegation

KtorZ committed:
adjust just-enough integration scenario w.r.t delegation

  Now that the wallet doesn't necessarily create a change output, some of those edgy scenarios needs adjustment. In particular, the wallet no longer needs to hold at least minUTxOValue to create a change output when delegating, having only the cost of delegation and key deposit amount is enough
